java-manta-client/src/main/java/com/joyent/manta/client/multipart/MantaMultipartManager.java

```diff
@@ -0,0 +1,1106 @@
+     *
+     * @param <R> Return type for executeWhenTimesToPollExceeded
+     * @param upload multipart upload object
+     * @param pingInterval interval to poll
+     * @param timesToPoll number of times to poll Manta to check for completion
+     * @param executeWhenTimesToPollExceeded lambda executed when timesToPoll has been exceeded
+     * @return null when under poll timeout, otherwise returns return value of executeWhenTimesToPollExceeded
+     * @throws IOException thrown if there is a problem connecting to Manta
+     */
+    public <R> R waitForCompletion(final MantaMultipartUpload upload,
+                                   final Duration pingInterval,
+                                   final int timesToPoll,
+                                   final Function<UUID, R> executeWhenTimesToPollExceeded)
+            throws IOException {
+        if (upload == null) {
+            throw new IllegalArgumentException("Upload must be present");
+        }
+
+        return waitForCompletion(upload.getId(), pingInterval, timesToPoll, executeWhenTimesToPollExceeded);
+    }
+
+    /**
+     * Waits for a multipart upload to complete. Polling for set interval.
+     *
+     * @param <R> Return type for executeWhenTimesToPollExceeded
+     * @param id multipart upload id
+     * @param pingInterval interval to poll
+     * @param timesToPoll number of times to poll Manta to check for completion
+     * @param executeWhenTimesToPollExceeded lambda executed when timesToPoll has been exceeded
+     * @return null when under poll timeout, otherwise returns return value of executeWhenTimesToPollExceeded
+     * @throws IOException thrown if there is a problem connecting to Manta
+     */
+    public <R> R waitForCompletion(final UUID id, final Duration pingInterval,
+                                   final int timesToPoll,
+                                   final Function<UUID, R> executeWhenTimesToPollExceeded)
+            throws IOException {
+        if (timesToPoll <= 0) {
+            String msg = String.format("times to poll should be set to a value greater than 1. "
+                    + "Actual value: %d", timesToPoll);
+            throw new IllegalArgumentException(msg);
+        }
+
+        final String dir = multipartUploadDir(id);
+        final MantaJob job = findJob(id);
+
+        if (job == null) {
+            String msg = "Unable for find job associated with multipart upload. "
+                    + "Was complete() run for upload or was it run so long ago "
+                    + "that we no longer have a record for it?";
+            MantaMultipartException e = new MantaMultipartException(msg);
+            e.setContextValue("upload_id", id.toString());
+            e.setContextValue("job_id", id.toString());
+
+            throw e;
+        }
+
+        MantaJobBuilder.Run run = mantaClient.jobBuilder().lookupJob(job);
+
+        if (LOG.isDebugEnabled()) {
```

Comment:
This guard is needed because we are doing a networking call to get information for the debug message. Without it, it would be executed with every invocation.